File: 02_Object_Oriented_Programming/02_Class_Relationships/04_Composition/python/house.py
"""
Composition — House class in Python.

A House creates and owns its Rooms internally (strong ownership).
Rooms cannot exist without the House.
"""

from room import Room


class House:
    """A house composed of rooms it creates and owns."""

    def __init__(self, address: str) -> None:
        self._address = address
        # Rooms are created INSIDE the House — composition
        self._rooms: list[Room] = [
            Room("Living Room", 350),
            Room("Kitchen", 200),
            Room("Bedroom", 300),
            Room("Bathroom", 100),
        ]

    @property
    def address(self) -> str:
        return self._address

    @property
    def rooms(self) -> list[Room]:
        return list(self._rooms)  # return copy

    @property
    def total_area(self) -> float:
        return sum(r.area_sq_ft for r in self._rooms)

    def show_house(self) -> None:
        print(f"House at {self._address} contains:")
        for r in self._rooms:
            print(f"  - {r}")
        print(f"  Total area: {self.total_area} sq ft")

    def __str__(self) -> str:
        return f"House({self._address}, rooms={len(self._rooms)})"

File: 02_Object_Oriented_Programming/02_Class_Relationships/04_Composition/javascript/house.js
/**
 * Composition — House class in JavaScript.
 *
 * A House creates and owns its Rooms internally (strong ownership).
 */

const Room = require("./room");

class House {
  #address;
  #rooms;

  constructor(address) {
    this.#address = address;
    // Rooms are created INSIDE the House — composition
    this.#rooms = [
      new Room("Living Room", 350),
      new Room("Kitchen", 200),
      new Room("Bedroom", 300),
      new Room("Bathroom", 100),
    ];
  }

  get address() {
    return this.#address;
  }

  get rooms() {
    return [...this.#rooms]; // return copy
  }

  get totalArea() {
    return this.#rooms.reduce((sum, r) => sum + r.areaSqFt, 0);
  }

  showHouse() {
    console.log(`House at ${this.#address} contains:`);
    for (const r of this.#rooms) {
      console.log(`  - ${r}`);
    }
    console.log(`  Total area: ${this.totalArea} sq ft`);
  }

  toString() {
    return `House(${this.#address}, rooms=${this.#rooms.length})`;
  }
}

module.exports = House;
